Stop resumable simulation driver at end of horizon

adrienmaillard opened the pull request:
* **Tickets addressed:** Fixes #859 
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->
Stop the resumable simulation driver (when simulating an activity) when the plan horizon is reached. 

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
A specific test has been added with a task that finishes after the end of the plan horizon.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->
None, this is a bugfix.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
Moaaar bugfixes

adrienmaillard committed:
Stop resumable simulation driver at end of horizon

Mythicaeda commented:
Just tested this with a goal that formerly would go off into forever. It now throws the following exception at the presumable end of scheduling:
```
[main] ERROR gov.nasa.jpl.aerie.scheduler.simulation.SimulationFacade - Simulation is probably outdated, check that no activity is removed between simulation and querying
java.util.NoSuchElementException: No value present
        at java.base/java.util.Optional.get(Unknown Source)
        at gov.nasa.jpl.aerie.scheduler.constraints.activities.ActivityCreationTemplate$1.valueAt(ActivityCreationTemplate.java:299)
        at gov.nasa.jpl.aerie.scheduler.constraints.activities.ActivityCreationTemplate$1.valueAt(ActivityCreationTemplate.java:265)
        at gov.nasa.jpl.aerie.scheduler.EquationSolvingAlgorithms$SecantDurationAlgorithm$1.valueAt(EquationSolvingAlgorithms.java:90)
        at gov.nasa.jpl.aerie.scheduler.EquationSolvingAlgorithms$SecantDurationAlgorithm.findRoot(EquationSolvingAlgorithms.java:113)
        at gov.nasa.jpl.aerie.scheduler.constraints.activities.ActivityCreationTemplate.rootFindingHelper(ActivityCreationTemplate.java:463)
        at gov.nasa.jpl.aerie.scheduler.constraints.activities.ActivityCreationTemplate.createInstanceForReal(ActivityCreationTemplate.java:323)
        at gov.nasa.jpl.aerie.scheduler.constraints.activities.ActivityCreationTemplate.createActivity(ActivityCreationTemplate.java:220)
        at gov.nasa.jpl.aerie.scheduler.solver.PrioritySolver.getBestNewActivities(PrioritySolver.java:716)
        at gov.nasa.jpl.aerie.scheduler.solver.PrioritySolver.satisfyGoalGeneral(PrioritySolver.java:538)
        at gov.nasa.jpl.aerie.scheduler.solver.PrioritySolver.satisfyGoal(PrioritySolver.java:376)
        at gov.nasa.jpl.aerie.scheduler.solver.PrioritySolver.solve(PrioritySolver.java:333)
        at gov.nasa.jpl.aerie.scheduler.solver.PrioritySolver.getNextSolution(PrioritySolver.java:122)
        at gov.nasa.jpl.aerie.scheduler.worker.services.SynchronousSchedulerAgent.schedule(SynchronousSchedulerAgent.java:210)
        at gov.nasa.jpl.aerie.scheduler.worker.SchedulerWorkerAppDriver.main(SchedulerWorkerAppDriver.java:90)
```

Adding the following unit test to SchedulingIntegrationTests will reproduce the exception:
```java
  @Test
  void testUnfinishedActivity(){
    final PlanningHorizon PLANNING_HORIZON = new PlanningHorizon(
        TimeUtility.fromDOY("2025-090T00:00:00"),
        TimeUtility.fromDOY("2025-134T00:00:00"));
    final var results = runScheduler(
        BANANANATION,
        List.of(),
        List.of(new SchedulingGoal(new GoalId(0L), """
          export default (): Goal => {
                           return Goal.ActivityRecurrenceGoal({
                             activityTemplate: ActivityTemplates.parent({
                               label: "label"
                           }),
                             interval: Temporal.Duration.from({ days: 30})
                           })
                       }
            """, true)), PLANNING_HORIZON);
  }
```

adrienmaillard commented:
Thank you Theresa ! As you told me on Slack, this is linked to this discussion https://github.com/NASA-AMMOS/aerie/pull/947#discussion_r1220430367
I fixed only one branch of the if statement code and not the other. I am gonna add a commit to this git branch to correct this behavior and factorize the code as I now remember @mattdailis suggested during a review.

adrienmaillard committed:
Factorize + protect rootfinding code (should have been done in #947)

adrienmaillard committed:
Fix insertion method

Would not insert activity in the future if removal list was empty

adrienmaillard committed:
Fix rootfinding failure case

If rootfinding fails to place an activity, the last tried activity placement is still in the sim and must be removed to avoid polluting the sim.

Mythicaeda commented on scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/SchedulingIntegrationTests.java:
```diff
@@ -1774,6 +1774,28 @@ export default function myGoal() {
+                           })
+                       }
+            """, true)), PLANNING_HORIZON);
+    //parent takes much more than 134 - 90 = 44 days to finish
+    assertEquals(0, results.updatedPlan.size());
```
Suggested addition:
```java
final var goalResult = results.scheduleResults.goalResults().get(new GoalId(0L));
assertFalse(goalResult.satisfied());
```

adrienmaillard replied:
Sure

Mythicaeda commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/activities/ActivityCreationTemplate.java:
```diff
@@ -286,41 +286,26 @@ public Duration valueAt(Duration start, HistoryWithActivity history) {
-            } catch (SimulationFacade.SimulationException e) {
-              logger.debug("Simulation error while trying to simulate activities: " + e);
+            } else{
+              logger.debug("No simulation error but activity duration could not be found in simulation, unfinished activity?");
```
Minor: Making this suggestion since we now know that unfinished activities will end up in this branch. 

```suggestion
              logger.debug("No simulation error but activity duration could not be found in simulation, likely caused by unfinished activity.");
```

adrienmaillard replied:
Yes

Mythicaeda reviewed (commented):
While this is passing test, I noticed a lot of errors in the output that I wasn't expecting.
Specifically, they were taking the form:

```
Task TaskId[id=948bf59e-efc7-4ede-8890-1b97073052a2] is ready but in unexpected execution state Terminated[startOffset=+00:00:00.000000, endOffset=+00:00:00.000000, joinOffset=+00:00:00.000000]
```
and
```
[Test worker] ERROR gov.nasa.jpl.aerie.scheduler.simulation.SimulationFacade - Simulation is probably outdated, check that no activity is removed between simulation and querying
```

Why are these now appearing?

adrienmaillard commented:
> While this is passing test, I noticed a lot of errors in the output that I wasn't expecting. Specifically, they were taking the form:
> 
> ```
> Task TaskId[id=948bf59e-efc7-4ede-8890-1b97073052a2] is ready but in unexpected execution state Terminated[startOffset=+00:00:00.000000, endOffset=+00:00:00.000000, joinOffset=+00:00:00.000000]
> ```
> 
> and
> 
> ```
> [Test worker] ERROR gov.nasa.jpl.aerie.scheduler.simulation.SimulationFacade - Simulation is probably outdated, check that no activity is removed between simulation and querying
> ```
> 
> Why are these now appearing?

For the first one... I don't know, seems it is stepping into a finished task which has no effect but is still weird I agree. 

For the second one, this is a message printed when the duration of an activity is queried but not found in the `ResumableSimulationDriver`. I think this message is misleading and I will remove it.

adrienmaillard committed:
Edit/remove log messages

Mythicaeda commented:
> seems it is stepping into a finished task which has no effect but is still weird

Is there a case where the current batch isn't updated after being processed? If so, that might be why we're seeing this message.

Mythicaeda reviewed (approved):
LGTM!

Regarding the `Task is in unexpected state` message: I checked against `develop` and I also see that message when running tests there. While we should look into why that's happening, fixing that is separate from this PR.

adrienmaillard commented:
> LGTM!
> 
> Regarding the `Task is in unexpected state` message: I checked against `develop` and I also see that message when running tests there. While we should look into why that's happening, fixing that is separate from this PR.

It happens just after a `simulateUntil` so maybe it has to do with how the batch is processed there.
